query-engine/query-engine-napi/src/lib.rs

```diff
@@ -0,0 +1,107 @@
+use engine::{ConnectParams, QueryEngine};
+use napi::{CallContext, Env, JsObject, JsString, JsUndefined, JsUnknown, Property};
+use napi_derive::{js_function, module_exports};
+use query_core::QueryExecutor;
+
+mod engine;
+mod error;
+mod exec_loader;
+
+pub(crate) type Result<T> = std::result::Result<T, error::ApiError>;
+pub(crate) type Executor = Box<dyn QueryExecutor + Send + Sync>;
+
+#[js_function(1)]
+fn constructor(ctx: CallContext) -> napi::Result<JsUndefined> {
+    let url = ctx.get::<JsString>(0)?.into_utf8()?;
+
+    let mut this: JsObject = ctx.this_unchecked();
+    let engine = QueryEngine::new(url.as_str()?)?;
+
+    ctx.env.wrap(&mut this, engine)?;
+    ctx.env.get_undefined()
+}
+
+#[js_function(1)]
+fn connect(ctx: CallContext) -> napi::Result<JsObject> {
+    let this: JsObject = ctx.this_unchecked();
+    let engine: &QueryEngine = ctx.env.unwrap(&this)?;
+    let engine: QueryEngine = engine.clone();
+
+    let arg0 = ctx.get::<JsUnknown>(0)?;
+    let params: ConnectParams = ctx.env.from_js_value(arg0)?;
+
+    ctx.env
+        .execute_tokio_future(async move { Ok(engine.connect(params).await?) }, |&mut env, ()| {
+            env.get_undefined()
+        })
+}
+
+#[js_function(1)]
+fn query(ctx: CallContext) -> napi::Result<JsObject> {
+    let this: JsObject = ctx.this_unchecked();
+    let engine: &QueryEngine = ctx.env.unwrap(&this)?;
+    let engine: QueryEngine = engine.clone();
+
+    let query = ctx.get::<JsObject>(0)?;
+    let body = ctx.env.from_js_value(query)?;
+
+    ctx.env
+        .execute_tokio_future(async move { Ok(engine.query(body).await?) }, |&mut env, response| {
+            env.to_js_value(&response)
```

Comment:
Convert big structs into JsObject has a lot of overhead, consider `serde_json::to_string` here and `JSON.parse` in  JavaScript side.

Reply:
Is this only objects with big arrays and large nested objects with lots of keys, or also objects with just a few keys and large values?

Reply:
I'm having the Chinese New Year, so I can't give specific performance data for now. I'll do some specific benchmark to see the specific performance data after the holiday.
But from my experience on the swc project, serde::to_string + JSON.parse seems always faster than the Env::to_js_value